File: backend/src/Services/RepairService.cs
using API.Entities;
using API.Interfaces;
using API.Types;

namespace API.Services;

public class RepairService(IRepairRepository repairRepository) {

    public async Task<Page<Repair>> GetAll(string filter, int page, int limit) {

        Page<Repair> repairs = await repairRepository.GetAll(filter, page, limit);

        return repairs;

    }

    public async Task<Page<Repair>> GetRepairsByApartment(int apartmentId, string filter, int page, int limit) {

        Page<Repair> repairs = await repairRepository.GetByApartmentId(apartmentId, filter, page, limit);

        return repairs;

    }

    public async Task<Repair?> GetRepairById(int id) {

        Repair? repair = await repairRepository.GetById(id);

        return repair;

    }

    public async Task<Repair?> CreateRepair(int userId, int apartmentId, string description, bool isRepaired) {

        Repair repair = new() {
            UserId = userId,
            ApartmentId = apartmentId,
            Description = description,
            IsRepaired = isRepaired
        };

        try {
            return await repairRepository.Create(repair);
        } catch {
            return null;
        }

    }

    public async Task<Repair?> UpdateRepair(int id, int? userId, int? apartmentId, string? description, bool? isRepaired) {

        Repair? repair = await repairRepository.GetById(id);

        if(repair == null) {
            return null;
        }

        repair.UserId = userId ?? repair.UserId;
        repair.ApartmentId = apartmentId ?? repair.ApartmentId;
        repair.Description = description ?? repair.Description;
        repair.IsRepaired = isRepaired ?? repair.IsRepaired;

        try {
            return await repairRepository.Update(repair);
        } catch {
            return null;
        }

    }

    public async Task<Repair?> DeleteRepair(int id) {

        Repair? repair = await repairRepository.GetById(id);

        if(repair == null) {
            return null;
        }

        try {
            return await repairRepository.Delete(repair);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/repair.service.ts
import { APIError } from "@/models/api-error.models";
import { Page } from "@/models/page";
import { Repair } from "@/models/repair.model";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class RepairService extends Service {

    static async getRepairsByApartment(apartmentId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Repair> | APIError>(`/api/repair/apartment/${apartmentId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createRepair(userId : number, apartmentId : number, description : string) {

        const response = await this.axios.post<Repair | APIError>("/api/repair", {
            userId,
            apartmentId,
            description
        });

        return response.data;

    }

    static async updateRepair(id : number, userId : optional<number>, apartmentId : optional<number>, description : optional<string>, isRepaired : optional<boolean>) {

        const response = await this.axios.put<Repair | APIError>(`/api/repair/${id}`, {
            userId,
            apartmentId,
            description,
            isRepaired
        });

        return response.data;

    }

    static async deleteRepair(id : number) {
        
        const response = await this.axios.delete<Repair | APIError>(`/api/repair/${id}`);

        return response.data;

    }

}